### ISSUE TITLE
ConstantIndex and Index not flagged as potential aliases

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#160525]
I tried this code ([playpen](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&code=fn+main%28%29+%7B%0A++++let+mut+arr+%3D+%5B%5BBox%3A%3Anew%2842%29%5D%5D%3B%0A++++let+alias+%3D+%26mut+arr%5B0%5D%5B%7B+let+%5Brow%5D+%3D+arr%3B+drop%28row%29%3B+0+%7D%5D%3B%0A++++println%21%28%22%7B%7D%22%2C+**alias%29%3B+%2F%2F+use-after-free+of+arr%27s+dead+stack+slot%0A%7D)):

```rust
fn main() {
    let mut arr = [[Box::new(42)]];
    let alias = &mut arr[0][{ let [row] = arr; drop(row); 0 }];
    println!("{}", **alias); // use-after-free of arr's dead stack slot
}
```

I expected to see this happen: Compiler rejects code as unsound.

Instead, this happened: Compiler accepts code and prints a garbage number. 

When run under Miri, we see this output:

```
error: Undefined Behavior: constructing invalid value of type std::boxed::Box<i32>: encountered a dangling box (use-after-free)
 --> src/main.rs:4:20
  |
4 |     println!("{}", **alias); // use-after-free of arr's dead stack slot
  |                    ^^^^^^^ Undefined Behavior occurred here
  |
  = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
  = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
```

Hypothesis: after a `move _1[ConstantIndex{0 of 1}]`, a subsequent use of `_1[Index(_6)]` is not flagged as a use-of-partially-moved. The "any subpath moved" check of borrowck treats `ConstantIndex` as a sibling of a runtime `Index` projection, rather than a potential alias.

### Meta

Nightly channel
Build using the Nightly version: 1.99.0-nightly

(2026-08-03 504869653f510b279c54)

**Comment by asquared31415** [comment]
Bisected:

```
searched toolchains nightly-2018-01-01 through nightly-2026-08-04


********************************************************************************
Regression in nightly-2019-04-23
********************************************************************************

looking for regression commit between 2019-04-22 and 2019-04-23
ending github query because we found starting sha: 31a75a172859d906d8e6a34af4afff9830af495c
get_commits_between returning commits, len: 6
  commit[0] 2019-04-21: Auto merge of #60124 - petrochenkov:stanomut, r=eddyb
  commit[1] 2019-04-21: Auto merge of #60158 - Xanewok:update-clippy, r=matthiaskrgr
  commit[2] 2019-04-22: Auto merge of #60053 - Xanewok:serde-save-analysis, r=nrc
  commit[3] 2019-04-22: Auto merge of #60133 - phansch:deny_rust_2018_idioms, r=Centril
  commit[4] 2019-04-22: Auto merge of #59114 - matthewjasper:enable-migate-2015, r=pnkfelix
  commit[5] 2019-04-22: Auto merge of #60168 - varkor:tidy-leading-newline, r=alexcrichton
```

This was previously rejected by the compiler in 2019-04-22:
```
error[E0382]: use of moved value: `arr[..][..]`
 --> src/main.rs:3:22
  |
3 |       let alias = &mut arr[0][{
  |  ______________________^
4 | |         let [row] = arr;
  | |              --- value moved here
5 | |         drop(row);
6 | |         0
7 | |     }];
  | |______^ value used here after move
  |
  = note: move occurs because `arr[..]` has type `[std::boxed::Box<i32>; 1]`, which does not implement the `Copy` trait
```

But began being accidentally accepted in 2019-04-23.

@rustbot label T-compiler I-unsound I-prioritize

**Comment by zachs18** [comment]
The bisection includes https://github.com/rust-lang/rust/pull/59114 which enabled NLL on edition 2015, which seems related; i.e. seems like NLL probably always had this issue.

**Comment by beepster4096** [comment]
I believe the `place_element_conflict: DISJOINT-OR-EQ-ARRAY-INDEX` case in `rustc_borrowck::places_conflict::place_projection_conflict` is incorrect. 

If an `Index` projection is being compared to any other array/slice access projection, there is an arbitrary level of overlap that cannot be determined. Instead of returning `Overlap::DisjointOrEqual`,  it should return `Overlap::Arbitrary`

**Comment by theemathas** [comment]
Out of curiosity, how was this bug discovered?

**Comment by carlini** [comment]
I sent this bug to @pnkfelix -- I'm testing a new LLM agent harness for finding soundness bugs in various programs. I have no idea what I'm doing with rust and so asked someone who knew something to actually check it and send it first. I have a bunch of other candidate bugs and I'm trying to teach myself enough about the rust borrow checker so that I can feel confident I'm not submitting slop that is /technically/ a bug but isn't something that's just not worth fixing. If you/anyone has thoughts on this happy to hear them.

**Comment by asquared31415** [comment]
Soundness bugs in general are worth reporting, because they're so important, as long as they're not related to misusing internal features (you'd have to explicitly opt into this, it would be obvious)

You might be able to work with people in the official zulip to verify bugs and/or ensure they're not duplicates.

**Comment by creeper6530** [comment]
> The bisection includes [#59114](https://github.com/rust-lang/rust/pull/59114) which enabled NLL on edition 2015, which seems related; i.e. seems like NLL probably always had this issue.

That's most probably true, since in nightly 2019-04-22 (the latest version before regression) setting the flag `-Zborrowck=migrate` also accepts the code:

```
~/tmp
❯ cat ./uaf.rs
fn main() {
    let mut arr = [[Box::new(42)]];
    let alias = &mut arr[0][{ let [row] = arr; drop(row); 0 }];
    println!("{}", **alias); // use-after-free of arr's dead stack slot
}

~/tmp
❯ rustup run nightly-2019-04-22 rustc uaf.rs
error[E0382]: use of moved value: `arr[..][..]`
 --> uaf.rs:3:22
  |
3 |     let alias = &mut arr[0][{ let [row] = arr; drop(row); 0 }];
  |                      ^^^^^^^^^^^^^^---^^^^^^^^^^^^^^^^^^^^^^^^
  |                      |             |
  |                      |             value moved here
  |                      value used here after move
  |
  = note: move occurs because `arr[..]` has type `[std::boxed::Box<i32>; 1]`, which does not implement the `Copy` trait

error: aborting due to previous error

For more information about this error, try `rustc --explain E0382`.

~/tmp
❯ rustup run nightly-2019-04-22 rustc uaf.rs -Zborrowck=migrate

~/tmp
❯ ./uaf
2055307368
```

**Comment by beepster4096** [comment]
> I believe the `place_element_conflict: DISJOINT-OR-EQ-ARRAY-INDEX` case in `rustc_borrowck::places_conflict::place_projection_conflict` is incorrect.
> 
> If an `Index` projection is being compared to any other array/slice access projection, there is an arbitrary level of overlap that cannot be determined. Instead of returning `Overlap::DisjointOrEqual`, it should return `Overlap::Arbitrary`

The code I blamed was added in #46268. However, the nightly immediately after (with -Zborrowck=mir) does not compile the code. I tracked the regression to 2018-01-11.

The only seemingly relevant PR would be #47167. It doesn't directly touch borrowck, but apparently it caused some code to start failing borrowck, so it could have caused code to stop failing borrowck too.

**Comment by beepster4096** [comment]
That PR caused the regression only because it fixed some indexing going through IndexMut incorrectly, masking the problem. I think I've barked up the wrong tree entirely and places_conflict is just fine.

**Comment by lqd** [comment]
`place_projection_conflict` will not be called in this example, so it's more like the OP said (one of the people who wrote the current borrowck): https://github.com/rust-lang/rust/blob/f73951df0a5566d94d13b7954acd9f4ab1fa3734/compiler/rustc_borrowck/src/lib.rs#L2135 missing that the `let [row] = arr;` is a partial move or something.

**Comment by beepster4096** [comment]
@rustbot claim

Forgot to do that, I've been working on a PR for this
